pkg/shell/authorized-keys.js

```diff
@@ -0,0 +1,228 @@
+                        .done(function(data) {
+                            var parts = data.split(" ");
+                            obj.size = parts[0];
+                            obj.fp = parts[1];
+                            obj.valid = true;
+
+                            /* if there is no comment, the filename
+                             * gets used, we don't want that
+                             */
+                            obj.comment = parts.slice(2).join(' ');
+                            obj.comment = obj.comment.replace(tmp_file_path, '').trim();
+                        })
+                        .always(function () {
+                            df.resolve(obj);
+                        });
+                })
+                .fail(function(ex) {
+                    df.reject(ex);
+                });
+            return df;
+        }
+
+        function parse_keys (content, tag, ex) {
+            var processed = 0;
+            var previous = { };
+            var seen = { };
+            var keys = [ ];
+            var tmp_file;
+            var tmp_file_path;
+            var lines = null;
+
+            last_tag = tag;
+
+            if (ex)
+                return process_failure(ex);
+
+            if (content && content.trim())
+                lines = content.trim().split('\n');
+
+            if (!lines)
+                return update_keys(keys, tag);
+
+            var i;
+            for (i = 0; i < self.keys.length; i++) {
+                previous[self.keys[i].raw] = self.keys[i];
+            }
+
+            function done_callback (obj) {
+                processed++;
+
+                if (!seen[obj.raw]) {
+                    seen[obj.raw] = true;
+                    keys.push(obj);
+                }
+
+                if (processed === lines.length) {
+                    tmp_file.close();
+                    cockpit.spawn(["rm", tmp_file_path]);
+                    update_keys(keys, tag);
+                } else {
+                    process_next();
+                }
+            }
+
+            function process_next() {
+                var raw = lines[processed];
+                if (previous[raw] === undefined) {
+                    create_key_object (raw, tmp_file, tmp_file_path)
+                        .done(done_callback)
+                        .fail(process_failure);
+                } else {
+                    done_callback(previous[raw]);
+                }
+            }
+
+            cockpit.spawn(["mktemp"])
+                .done(function(result) {
+                    tmp_file_path = result.trim();
+                    tmp_file = cockpit.file(tmp_file_path);
+                    process_next();
+                })
+                .fail(process_failure);
+        }
+
+        self.add_key = function(key) {
+            var df = $.Deferred();
+            key = key.trim();
+            cockpit.spawn(["mktemp"])
```

Comment:
Why are we doing so many round trips here. I would really suggest just passing in a shell script, with the key as input, and performing all these actions from the script.